src/main/java/org/opensearch/security/configuration/SecurityConfigVersionsLoader.java

```diff
@@ -0,0 +1,192 @@
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import org.opensearch.action.get.GetRequest;
+import org.opensearch.action.get.GetResponse;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.security.DefaultObjectMapper;
+import org.opensearch.security.support.ConfigConstants;
+import org.opensearch.transport.client.Client;
+
+/**
+ * Utility class to load security configuration version documents
+ */
+
+public class SecurityConfigVersionsLoader {
+    private static final Logger log = LogManager.getLogger(SecurityConfigVersionsLoader.class);
+
+    private final Client client;
+    private final String opendistroSecurityConfigVersionsIndex;
+
+    public SecurityConfigVersionsLoader(Client client, Settings settings) {
+        this.client = client;
+        this.opendistroSecurityConfigVersionsIndex = settings.get(
+            ConfigConstants.SECURITY_CONFIG_VERSIONS_INDEX_NAME,
+            ConfigConstants.OPENDISTRO_SECURITY_DEFAULT_CONFIG_VERSIONS_INDEX
+        );
+    }
+
+    private void getSecurityConfigVersionDocAsync(ActionListener<SecurityConfigVersionDocument> listener) {
+        GetRequest getRequest = new GetRequest(opendistroSecurityConfigVersionsIndex, "opendistro_security_config_versions");
+
+        client.get(getRequest, new ActionListener<>() {
+            @Override
+            public void onResponse(GetResponse getResponse) {
+                try {
+                    if (!getResponse.isExists()) {
+                        log.warn("Config versions document not found in {}", opendistroSecurityConfigVersionsIndex);
+                        listener.onResponse(new SecurityConfigVersionDocument()); // return empty doc
+                        return;
+                    }
+
+                    SecurityConfigVersionDocument doc = DefaultObjectMapper.readValue(
+                        getResponse.getSourceAsString(),
+                        SecurityConfigVersionDocument.class
+                    );
+
+                    doc.setSeqNo(getResponse.getSeqNo());
+                    doc.setPrimaryTerm(getResponse.getPrimaryTerm());
+
+                    listener.onResponse(doc);
+                } catch (IOException e) {
+                    log.error("Failed to parse config versions doc", e);
+                    listener.onFailure(e);
+                }
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                log.error("Failed to load config versions doc from {}", opendistroSecurityConfigVersionsIndex, e);
+                listener.onFailure(e);
+            }
+        });
+    }
+
+    public void loadLatestVersionAsync(ActionListener<SecurityConfigVersionDocument.Version<?>> listener) {
```

Comment:
If the list is always ordered then can we reliably get the latest by getting the last in the list?